### PR TITLE
Metabolizing property

### PR DESCRIPTION
**Pull request by Arsakes**
Pull contains working implementation of metabolizng property for both plants and characters. Character metabolise function was deleted and its functionality moved to the property. 
Metabolizing property relies on following new numercial properties:
- massreserves
- reservelimit
- mass
- status
- grows

Growth code from `Plant.cpp` was moved to metabolizing property aswell.
Currently entites that are alive have two kinds of mass:
- mass - total mass of entity
- massserseves - biomass that can be changed into energy without triggering death.

You can think of it as 'fat' in case of animals, fruit mass in case of trees etc.
reservelimit - is maximal mass reservers of an entity. It is normalized value a number between 0 (entity cannot store energy as mass) and 1 (entity can be made entirely of fat).

Things that require attention:
1. Writing proper integration tests.
2. Adding `<metabolizing></metabolizing>` property to entity classes that should be able to metabolize.

### CODE REVIEW

File: rulesets/MetabolizingProperty.cpp [hunk 1]
```diff
@@ -0,0 +1,407 @@
+// Cyphesis Online RPG Server and AI Engine
+// Copyright (C) 2013 Alistair Riddoch
```
Comment (erikogenvik) [inline rulesets/MetabolizingProperty.cpp]: You should probably add yourself as an author here too.

Now, for most files that I alter I don't do this (Git contains the record anyway), but if it's a new file you should have your name there too.

**Comment by erikogenvik** [comment]
I'll take a closer look, however at the moment I'm trying to focus on getting physics in there.

I would however want to see some tests for metabolizing sanity. I.e. that things grow to max size, wither and die in accordance to the rules. A test where you send 1000 Ticks to a metabolizing entity and then check that the rules are followed.

This is because this is one of those tricky game logic things where errors can be both hard to catch (are things really growing as they should) and have large consequences (all trees have died).

**Comment by erikogenvik** [comment]
I've tried out this branch now.
First thing that sticks out it that characters don't metabolize. I tested this by spawning a standard human character and eating a ham. Status is unchanged, even after a couple of minutes.

Also, am I right to assume that with these patches plants won't metabolize at all, as there's no default metabolizing property on them?

And I still would really want to see proper tests for this property. I.e. "what happens after 100 ticks" and so on.

**Comment by Arsakes** [comment]
I wasn't sure to which parent class attach "metabolizing" propety therefore characters don't metabolize yet. Should I attach the property to mobile class? From what I've seen its shared for all animals.
The other class possessing metabolizing property would be plant - would that be right?

I'll get you know after those tests will be ready.

**Comment by erikogenvik** [comment]
I took a look at this again, but the code doesn't compile.
I got some errors in Plant.cpp.
Also note that we now use CMake, so you need to add your files to the CMakeLists.txt file and remove the Makefile.am files.

**Comment by Arsakes** [comment]
Thank you, also I need to change some in relation to mass being readonly.
My idea is to split mass changes into "size change" (aka volume) and  (density).

**Comment by Arsakes** [comment]
Ok, this also would also implicate that storage of energy via getting fat would not increase the real mass (since this would have to change the entity size).
But I assume everyone is ok with that (its certainly simpler).

**Comment by erikogenvik** [comment]
Could you walk me through the difference between getting fat and growing,
in the current system? I haven't looked at the code in a while.

/Erik

2017-12-18 13:50 GMT+01:00 pifthemagicdragon <notifications@github.com>:

**Comment by Arsakes** [comment]
In current code (that as you pointed out in e-mail a while ago - is wrong):
Growing is increasing mass due to volume increase.
Getting fat is increase of mass that cannot be accounted by or associated with size increase.
The way I allow both of them to happen at the same time is by keeping track of two variables:

    mass  // total mass
    massReserve  // mass of fat
 
Whenever there's  some food consumed and status is high enough so both fat and size should grow.
The growth is calculated according to forumals:
            
    double coreMass = mass-massReserve;
    ...
    // Getting bigger (growth) is a size multiplier
    growth = 1.0f + (1.0f - 0.5*reserveLimit)*massChange/coreMass; 
    // Getting fatter
    massReserve += reserveLimit * massChange; 
    // Mass
    mass += massChange;

reserveLimit is a constant

However this is wrong because it doesn't play nicely with density (and I'm not suppose to modify mass).
I hope that helps.

Now when I look at this I agree that this unnecessary complexity by trying to be physically accurate.

In this case the price for simplicity is having fat entities and skinny entities (of the same type) having the same mass.
If you OK with that I am too I just want to finish it.

**Pull request closed without merging**